File: c/reverse-string/reverse_string.c
#include <string.h>
#include <malloc.h>
#include "reverse_string.h"

char *reverse(const char *str)
{
	size_t len = strlen(str);

	char *reversed = malloc(len + 1);
	if (!reversed)
		return NULL;

	for (size_t i = 0; i < len; i++)
		reversed[i] = str[len - i - 1];

	reversed[len] = '\0';

	return reversed;
}

File: zig/reverse-string/reverse_string.zig
const std = @import("std");

/// Writes a reversed copy of `s` to `buffer`.
pub fn reverse(buffer: []u8, s: []const u8) []u8 {
    const dest = buffer[0..s.len];

    for (0..s.len) |i|
        buffer[i] = s[s.len - i - 1];

    return dest;
}
